### DISCUSSION TITLE
Playwright-Artillery Test-integration

### DISCUSSION THREAD

**Discussion by ChetanKB** [artilleryio/artillery#1473] [Artillery]
Hi all,
I am Trying to Execute the test.spec.js file Via the artillery, the test.spec.js contains all the tests which I want to execute via the artillery engine but to execute javascript, I need to mention the function name in the YAML file But my Javascript doesn't have any function name so is there any way available to execute the multiple tests via artillery engine.

### Here is The code of my spec.js :

```
const { test, expect } = require('@playwright/test');

let d = new Date();
let timestamp = d.getTime();

test.describe.parallel('group', () => {

  test('test-1', async ({ page }) => {


    await page.goto('https://en.wikipedia.org/wiki/Main_Page');

    await page.locator('[placeholder="Search Wikipedia1"]').click();

    await page.locator('[placeholder="Search Wikipedia"]').fill('testing');
    // Press Enter
    await Promise.all([
      page.waitForNavigation(),
      page.locator('[placeholder="Search Wikipedia"]').press('Enter')
    ]);
   // await page.screenshot({ path: './screenshots/test-1-' + timestamp + '.png' });
    // Go to https://en.wikipedia.org/wiki/Main_Page
    await page.goto('https://en.wikipedia.org/wiki/Main_Page');

  });

  test('test-2', async ({ page }) => {

    
  await page.goto('https://en.wikipedia.org/wiki/Main_Page');

  await page.waitForLoadState();

  await page.locator('[placeholder="Search Wikipedia"]').click();

  await page.locator('[placeholder="Search Wikipedia"]').fill('automation');

  await Promise.all([
      page.waitForNavigation(),
      page.locator('[placeholder="Search Wikipedia"]').press('Enter')
    ]);
//await page.screenshot({ path: './screenshots/test-2-' + timestamp + '.png' });

  await page.locator('text=Main page').click();
  await expect(page).toHaveURL('https://en.wikipedia.org/wiki/Main_Page');


  });
})
```

### This is the Project Structure:

![issue](https://user-images.githubusercontent.com/53107584/172354113-274edb0b-8e0e-4a10-a6d7-79a62c409bd9.png)

**Comment by hassy** [comment]
Hi @ChetanKB 👋 

Right now Artillery needs a function that wraps Playwright actions. It's pretty easy to do though and only requires a simple refactoring of existing code:

1. Extract Playwright steps into their own function
2. Call that function from your existing `test()` steps
3. Create a wrapper function for Artillery for load testing

For example, extract existing Playwright steps into their own function:

```js
async function testWikipedia(page) {
    await page.goto('https://en.wikipedia.org/wiki/Main_Page');

    await page.locator('[placeholder="Search Wikipedia1"]').click();

    await page.locator('[placeholder="Search Wikipedia"]').fill('testing');
    // Press Enter
    await Promise.all([
      page.waitForNavigation(),
      page.locator('[placeholder="Search Wikipedia"]').press('Enter')
    ]);
   // await page.screenshot({ path: './screenshots/test-1-' + timestamp + '.png' });
    // Go to https://en.wikipedia.org/wiki/Main_Page
    await page.goto('https://en.wikipedia.org/wiki/Main_Page');
}
```

Use that function in your existing tests:

```js
test.describe.parallel('group', () => {
  test('test-1', async ({ page }) => {
    await testWikipedia(page);
  });
});
```

Wrap it in a function for Artillery to call:

```js
async function artilleryScenario(page) {
  await testWikipedia();
}

module.exports = { artilleryScenario };
```

And then in the Artillery script use that function:

```yaml
config:
  processor: ./my-js-code.js
scenarios:
  - engine: playwright
    flow: []
    flowFunction: artilleryScenario
```

**Reply by ChetanKB** [reply]
Hi @hassy 
Whatever you mentioned above that's working fine and I am able to execute multiple multiple tests, But  Is there any way we need to check whether the Test is passed or Failed after the execution.

**Reply by hassy** [reply]
hi @ChetanKB 👋 

> check whether the Test is passed or Failed after the execution.

What would you expect that check to look like? i.e. what would be a pass/fail result for the load tests you're running with Playwright + Artillery?

If you're looking for a pass/fail condition based on one of the metrics generated by the test run, `ensure` is a good fit: https://www.artillery.io/docs/guides/guides/test-script-reference#ensure---slo-checks

**Reply by afsal-backer** [reply]
Hi @hassy  , I got this working with your refactoring suggestion. My package.json has

    "scripts": {
        "build": "tsc -p tsconfig.json",
        "art": "$(npm bin)/artillery run ./playwright-art/artillery-scenarios/presser-load-test.yml",
        "play": "playwright test  ./playwright-art/src/areas/presser/presser.spec.ts --headed"
    },
and I have to run `npm run art`  to run Artillery test and `npm run play` to run Playwright test.

Problem I see with running UI test in artillery is that I do not get Playwright test pass/fail message/report like shown below (which is what I think @ChetanKB  was referring to above)

```
Running 1 test using 1 worker
  ✓  1 src/areas/presser/presser.spec.ts:6:9 › presser End to End tests › I should be able to view the presser (1.6s)
  1 passed (2.4s)
```

It is crucial to know if the test failed because of a website issue. Is there a way to get that kinda reporting when running Artillery tests?

Or, is there a way to integrate both as 1 single execution? Like - driving the test with Playwright and doing load scenarios with Artillery?

**Comment by rakesh-rameshtr** [comment]
@hassy How do have handle custom page object ? 
I want to use existing session storage so that user is always logged in when the page is loaded. I accomplish that in playwright using the following command .  Is there a way to do this in artilliery playwright ? Otherwise i need to include a method to login each time when the test is executed

` 

     async function loadTest(page)
     {
    
    const browser = await chromium.launch();
    const sellerPage = await (
        await browser.newContext({
            storageState: "StorageState/storageState_user.json"
        })
    ).newPage();
    }

**Reply by hassy** [reply]
That's not possible at the moment @rakesh-rameshtr but we will can the ability to extend the browser context in the Playwright engine.

**Reply by hassy** [reply]
@rakesh-rameshtr it's now possible to customize the browser context in the most recent version of the Artillery Playwright integration. You can check out the examples here: https://github.com/artilleryio/artillery-engine-playwright#configuration

**Reply by srivishal1** [reply]
Hi Hassy,[ChetanKB](https://github.com/ChetanKB)

I wrote the same code but I am getting below error-

Error: flows.js:6:7: Playwright Test did not expect test.describe() to be called here.
Most common reasons include:
- You are calling test.describe() in a configuration file.
- You are calling test.describe() in a file that is imported by the configuration file.
- You have two different versions of @playwright/test.This usually happens
  when one of the dependencies in your package.json depends on @playwright/test.

My package json looks like - 
{
  "dependencies": {
    "artillery": "^2.0.0-24",
    "artillery-engine-playwright": "^0.2.1"
  },
  "devDependencies": {
    "@playwright/test": "^1.28.1"
  }
}

**Reply by hassy** [reply]
you need to refactor your code as described in the comment above: https://github.com/artilleryio/artillery/discussions/1473#discussioncomment-2898343
